### PR TITLE
[#89] Write analysis for concept exercise `need-for-speed`

### PR DESCRIPTION
**Pull request by jiegillet**
I hit a wall with this one and I need some help.

I think I found a bug related to `feature` matching multiline strings and string interpolation... Maybe?

I added some tests in `test/elixir_analyzer/exercise_test/feature/string_interpolation_test.exs` the first 3 pass and the next ones fail. I'm not sure where to begin.

### CODE REVIEW

**Comment by angelikatyborska** [comment]
>  I'm not sure where to begin.

Those are general hints that could make debugging easier but I don't know really what's going on:

I would start by making your intentions more clear by specifying a `find` value in the `feature` calls. 

Then, for every test that you want to pass, I would add an opposite test. This is to make sure that the check actually works and isn't ignored. E.g.

for
```elixir
  test_exercise_analysis "normal string",
    comments_exclude: ["normal string"] do
    [
      defmodule MyModule do
        def hi(name) do
          "hello you"
        end
      end
    ]
  end
```
add
```elixir
  test_exercise_analysis "not a normal string",
    comments_include: ["normal string"] do
    [
      defmodule MyModule do
        def hi(name) do
          "hello you!!!!!!"
        end
      end
    ]
  end
```

Then, I would comment out all the failing tests but the first one and their corresponding `feature` calls so that you can focus on a single weird case.

Another tip, it's impossible to tell apart multiline strings from normal strings by the AST so I would drop that distinction (or make it clear in the tests that you know it's the same thing):

```
iex(4)> quote do
...(4)> """
...(4)> hello
...(4)> """
...(4)> end
"hello\n"

iex(5)> quote do
...(5)> "hello\n"
...(5)> end
"hello\n"
```


I think it is possible that the `feature` macro is unable to handle blocks that are _just_ strings with interpolation and no other calls 🤔 but I don't know why atm...

**Comment by neenjaw** [comment]
Had a spot of internet, I think it is because `"""` is a macro in and of itself.  The current match algorithm depend on the quoted ast of the form to expand all at once, so that we can replace things (lib/elixir_analyzer/exercise_test/feature.ex Line 75-122) and create a pattern that doesn't expand any further when used (lib/elixir_analyzer/exercise_test/feature/compiler.ex Line 48-65).

At least in the case of string interpolation in the context of a multiline string, this seems to fail:

```
iex(2)> quote do  
...(2)>   """     
...(2)>   #{value}\n
...(2)>   """       
...(2)> end         
{:<<>>, [],
 [
   "",
   {:"::", [],
    [
      {{:., [], [Kernel, :to_string]}, [], [{:value, [], Elixir}]},
      {:binary, [], Elixir}
    ]},
   "\n\n"
 ]}
```

This is the quoted form of the code segment.

```
"{:<<>>, _, [\"hello \", {:\"::\", _, [{{:., _, [Kernel, :to_string]}, _, [{:name, _, nil}]}, {:binary, _, nil}]}]}"
```
This is the string pattern created by the feature macro on `lib/elixir_analyzer/exercise_test/feature.ex:L115`

```
{:{}, [line: 1],
 [
   :<<>>,
   {:_, [line: 1], nil},
   [
     "hello ",
     {:{}, [line: 1],
      [
        :"::",
        {:_, [line: 1], nil},
        [
          {:{}, [line: 1],
           [
             {:{}, [line: 1],
              [
                :.,
                {:_, [line: 1], nil},
                [{:__aliases__, [line: 1], [:Kernel]}, :to_string]
              ]},
             {:_, [line: 1], nil},
             [{:{}, [line: 1], [:name, {:_, [line: 1], nil}, nil]}]
           ]},
          {:{}, [line: 1], [:binary, {:_, [line: 1], nil}, nil]}
        ]
      ]},
     "\n"
   ]
 ]}
```
This is the evaluated string created by the feature compiler on `lib/elixir_analyzer/exercise_test/feature/compiler.ex:L52`. Notice the metadata reinserted into the ast.  This will never match anything because of that addition.

I don't know of any Elixir examples of creating custom pattern matchers. And as far as I know it is otherwise impossible to create dynamic patterns to leverage `=/2` or `match?/2`.  Interestingly, ETS might be an example we could follow to create out own matcher.  See this SO post by rvirding: https://stackoverflow.com/questions/3989271/dynamic-pattern-matching


---

TLDR: Right now we can't do this check, but we can pull an ETS and build a custom matcher... maybe

**Comment by jiegillet** [comment]
I see. I saw that pattern when I was debugging around, but did not know that it wasn't supposed to be that way. I guess I'll clean up the tests and keep familiarizing myself with the codebase and stuff :)

> Had a spot of internet

I hope I'm not ruining your vacation :D

**Comment by neenjaw** [comment]
> I hope I'm not ruining your vacation

Nope, just thought it worthwhile to share what I found about why it wasn't working to save some time and headache.

**Comment by angelikatyborska** [comment]
@jiegillet are you stuck? do you need more help?

**Comment by jiegillet** [comment]
I was stuck for a while, so I took a break but kept thinking about it. I've started looking at it again yesterday, and I'm making progress understanding the issue. I think it is a bit different from what we discussed before, but I need a bit more time to be sure :)

**Comment by jiegillet** [comment]
OK, I finally came up with something. This is possibly a big change so I will justify it a bit and I would love to have your input.
Here what I think was going wrong

```elixir
iex(85)> quote do
...(85)>   def hi(name) do
...(85)>     "hello #{name}\n"
...(85)>   end
...(85)> end
{:def, [context: Elixir, import: Kernel],
 [
   {:hi, [context: Elixir], [{:name, [], Elixir}]},
   [
     do: {:<<>>, [],
      [
        "hello ",
        {:"::", [],
         [
           {{:., [], [Kernel, :to_string]}, [], [{:name, [], Elixir}]},
           {:binary, [], Elixir}
         ]},
        "\n"            # Note this single newline...
      ]}
   ]
 ]}
iex(86)> |> Macro.to_string
"def(hi(name)) do\n  \"hello \#{name}\n  \"\nend"
iex(87)> |> tap(&IO.puts/1)
def(hi(name)) do
  "hello #{name}
  "                      # ... that becomes an actual newline instead of being escaped
end
"def(hi(name)) do\n  \"hello \#{name}\n  \"\nend"
iex(88)> |> Code.string_to_quoted!
{:def, [line: 1],
 [
   {:hi, [line: 1], [{:name, [line: 1], nil}]},
   [
     do: {:<<>>, [line: 2],
      [
        "hello ",
        {:"::", [line: 2],
         [
           {{:., [line: 2], [Kernel, :to_string]}, [line: 2],
            [{:name, [line: 2], nil}]},
           {:binary, [line: 2], nil}
         ]},
        "\n  "     # and drags the indentation with it
      ]}
   ]
 ]}
```

Basically `Macro.to_string` confuses the ending newline in the string for an actual newline, write the ending quote indented and the indentation gets dragged in by `Code.string_to_quoted!`. This may be something similar to [this issue](https://github.com/elixir-lang/elixir/issues/5318)? 

So I decided to try not going through the intermediate string and match the AST directly, by replacing `match?` by a custom `form_match?`. It seems to work fine. Some side effects I could think of:
* Good side effect: I think this can solve #93 by letting people use `form` within `form` and using `form_match?` deal with it
* Bad side effect: I haven't checked yet, but this probably breaks #117

**Comment by angelikatyborska** [comment]
🤯 Great detective work!

This is definitely a bug in Elixir. I downloaded 1.12.2 to make sure it's still there, it is. I think we should open an issue in the Elixir GitHub repo. Do you want to do the honors, or should I? I looked through `Macro.to_string` related issues but there are no open issues for this specific problem.

So basically for this input:

```
iex()> quoted =
...()> quote do
...()>   def hi(name) do
...()>     "hello #{name}\n"
...()>   end
...()> end
{:def, [context: Elixir, import: Kernel],
 [
   {:hi, [context: Elixir], [{:name, [], Elixir}]},
   [
     do: {:<<>>, [],
      [
        "hello ",
        {:"::", [],
         [
           {{:., [], [Kernel, :to_string]}, [], [{:name, [], Elixir}]},
           {:binary, [], Elixir}
         ]},
        "\n"
      ]}
   ]
 ]}

iex()> string = Macro.to_string(quoted)
```

Actual output
```elixir
"def(hi(name)) do\n  \"hello \#{name}\n  \"\nend"
```
```elixir
# IO.puts(string)
def(hi(name)) do
  "hello #{name}
  "
end
```

Expected output
```elixir
"def(hi(name)) do\n  \"hello \#{name}\\n\"\nend"
```
```elixir
def(hi(name)) do
  "hello #{name}\n"
end
```

> Bad side effect: I haven't checked yet, but this probably breaks

It does 🤔 I'll look into it, maybe it's possible to fix?

**Comment by angelikatyborska** [comment]
@jiegillet Here's my proposal on how to fix #117 and keep your changes, in a commit: https://github.com/exercism/elixir-analyzer/commit/b27e59d4d990e1a402723e385904a5519ea4b9e2

The real input to the analyzer will be a string. So for tests where this weird bug changes the behavior, I propose we also define the test data as a string.

**Comment by jiegillet** [comment]
> This is definitely a bug in Elixir. I downloaded 1.12.2 to make sure it's still there, it is. I think we should open an issue in the Elixir GitHub repo. Do you want to do the honors, or should I? I looked through `Macro.to_string` related issues but there are no open issues for this specific problem.

Alright, [done](https://github.com/elixir-lang/elixir/issues/11172). Thank you for your help!

> The real input to the analyzer will be a string. So for tests where this weird bug changes the behavior, I propose we also define the test data as a string.

Sure, that makes sense. I'm glad you were able to align #117 so quickly.
What's the next step? Reviews from @neenjaw? 
I don't really want to write new analyses if the AST matching is going to change.

**Comment by angelikatyborska** [comment]
From your side, you could merge main to this branch to resolve conflicts, and then you can cherry-pick my fix into this branch too (https://github.com/exercism/elixir-analyzer/commit/b27e59d4d990e1a402723e385904a5519ea4b9e2) because the changes are all for code from this branch, not from #117. I think there are dialyzer warnings, if that's true they might be unsolvable and might need to be ignored, e.g. like here https://github.com/exercism/elixir-analyzer/blob/b27e59d4d990e1a402723e385904a5519ea4b9e2/lib/elixir_analyzer/exercise_test.ex#L20

I'll ping Tim to see if he's back and interested in a review. If not, I'll approve and merge 🙂

**Comment by jiegillet** [comment]
Well, the issue is already closed
> This has been fixed in Elixir master since Macro.to_string has been rewritten to use the same code path as the code formatter. Thank you!

**Review by angelikatyborska — APPROVED** [review]

**Comment by angelikatyborska** [comment]
Tim says we can merge without his review 🙂
